Feature/go synchronizer: TIL and db tests

kaiden-araki opened the pull request:
This PR updates the code for TIL and db tests.

- Modified TIL to make appropriate for SYB (`createVouch`, `deleteVouch`, and `exit` )
- Initial setup for batch testing
- Tx testing with PoolL2 on l2db
- Resolve conflicts with go-sequencer branch from @Shailu-s

Shailu-s commented:
Hey @kaiden-araki , I’ve reviewed the PR it looks good and I've approved it. 

However, there are a few points which we can discuss and consider updating going forward:

Since we’ll need to manage and update two separate Merkle trees in our StateDB, I suggest moving the helper functionality from `go_sequencer/common/account.go` into a new module. This module could have the existing helpers like `Bytes`, `HashValue`, and `BigInts`, along with the new ones we’ll add for links. We could also place the Account and Link structs there.

Initially, I considered having separate modules for account and link, but we'll need to check for dependency issues if we go that route.

Also, to have some test the implementation, I had placed some of these functions in `go_sequencer/database/statedb/account_service.go` and `go_sequencer/database/statedb/link_service.go` (e.g., `PutAccount`, `PutLink`, etc.), which handle storing accounts and links in the Merkle tree, converting data to bytes, calculating Poseidon hashes, etc. 

Now, Since these updates are now in place, we can comment it out for reference until everything is finalized, OR could also just remove them.

Additionally, I added some dummy struct models for accounts and links in `go_sequencer/models/account.go` and `go_sequencer/models/link.go`, respectively. These can also be removed once we finalize the structure.

Let me know your thoughts on all these same and We can also discuss and plan to have these updates moving further as per our needs. 

Thanks :pray:

Shailu-s reviewed (approved):
Hey @kaiden-araki , I’ve reviewed the PR it looks good and I've approved it. 

However, there are a few points which we can discuss and consider updating going forward:

Since we’ll need to manage and update two separate Merkle trees in our StateDB, I suggest moving the helper functionality from `go_sequencer/common/account.go` into a new module. This module could have the existing helpers like `Bytes`, `HashValue`, and `BigInts`, along with the new ones we’ll add for links. We could also place the Account and Link structs there.

Initially, I considered having separate modules for account and link, but we'll need to check for dependency issues if we go that route.

Also, to have some test the implementation, I had placed some of these functions in `go_sequencer/database/statedb/account_service.go` and `go_sequencer/database/statedb/link_service.go` (e.g., `PutAccount`, `PutLink`, etc.), which handle storing accounts and links in the Merkle tree, converting data to bytes, calculating Poseidon hashes, etc. 

Now, Since these updates are now in place, we can comment it out for reference until everything is finalized, OR could also just remove them.

Additionally, I added some dummy struct models for accounts and links in `go_sequencer/models/account.go` and `go_sequencer/models/link.go`, respectively. These can also be removed once we finalize the structure.

Let me know your thoughts on all these same and We can also discuss and plan to have these updates moving further as per our needs. 

Thanks :pray:

kaiden-araki commented:
Hey @Shailu-s,
Here, I have updated structure for account and vouch which represents several merkle trees in stateDB.

- Defined the struct and common funcs for account and vouch in `common` module.
- Added KV storage operations in `stateDB` module.
- Tested vouch with merkle tree inside stateDB.

kaiden-araki commented on go_sequencer/common/account.go:
```diff
@@ -1,67 +1,195 @@
-	// NLeafElems is the number of elements for a leaf
-	NLeafElems = 4
+	// NAccountLeafElems is the number of elements for a leaf in account tree
+	NAccountLeafElems = 4
```
I think it is needed, because we have leaves seperated by `NAccountLeafElems`.

Of course, it can be hardcoded as like 128 rather than `32 * NAccountLeafElems`, but we need to make sure that those leaves are seperated by finite filed of 32 bytes, and the total number of elements of leaf is 4, that's the reason we have `NAccountLeafElems` here.

Additionally, It is more readable and understandable.

kaiden-araki commented on go_sequencer/common/account.go:
```diff
@@ -1,67 +1,195 @@
+// is represented by 32 bytes, in spite of the BigInt could be represented in
+// less bytes (due a small big.Int), so in this way each BigInt is always 32
+// bytes and can be automatically parsed from a byte array.
+func (a *Account) Bytes() ([32 * NAccountLeafElems]byte, error) {
```
Same as above

kaiden-araki commented on go_sequencer/common/account.go:
```diff
@@ -1,67 +1,195 @@
+		ayBytes[0] = ayBytes[0] & 0x3f //nolint:gomnd
+		pkY = big.NewInt(0).SetBytes(ayBytes)
+	}
+	finiteFieldMod, ok := big.NewInt(0).SetString("21888242871839275222246405745257275088548364400416034343698204186575808495617", 10) //nolint:gomnd
```
In elliptic curve, pkY should be inside the finiteField so that we can use it for further purpose.
And regarding the size of pkY, it is too long to store, so we need to mod pkY by finiteFieldMod, which is 32 bytes.

kaiden-araki commented on go_sequencer/common/vouch.go:
That's right.
And we can follow above logic to handle and store the vouchIdx as the type of integer.

Thanks
